pkg/adaptor/cloud/ibmcloud/powervs_provider.go

```diff
@@ -0,0 +1,227 @@
+// Copyright Confidential Containers Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package ibmcloud
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"log"
+	"net"
+	"time"
+
+	"github.com/IBM-Cloud/power-go-client/power/models"
+	"github.com/IBM/go-sdk-core/v5/core"
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/cloud"
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/util"
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/util/cloudinit"
+)
+
+const (
+	retries             = 25
+	pollingInterval     = 5
+	dhcpPollingInterval = 30
+)
+
+var pvslogger = log.New(log.Writer(), "[adaptor/cloud/ibmcloud-powervs] ", log.LstdFlags|log.Lmsgprefix)
+
+type ibmcloudPowerVSProvider struct {
+	powervsService
+	serviceConfig *PowerVSConfig
+}
+
+func NewPowerVSProvider(config *PowerVSConfig) (cloud.Provider, error) {
+
+	pvslogger.Printf("ibmcloud-powervs config: %#v", config.Redact())
+
+	powervs, err := newPowervsClient(config.ApiKey, config.ServiceInstanceID, config.Zone)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ibmcloudPowerVSProvider{
+		powervsService: *powervs,
+		serviceConfig:  config,
+	}, nil
+}
+
+func (p *ibmcloudPowerVSProvider) CreateInstance(ctx context.Context, podName, sandboxID string, cloudConfig cloudinit.CloudConfigGenerator) (*cloud.Instance, error) {
+
+	instanceName := util.GenerateInstanceName(podName, sandboxID, maxInstanceNameLen)
+
+	userData, err := cloudConfig.Generate()
+	if err != nil {
+		return nil, err
+	}
+
+	body := &models.PVMInstanceCreate{
+		ServerName:  &instanceName,
+		ImageID:     &p.serviceConfig.ImageID,
+		KeyPairName: p.serviceConfig.SSHKey,
+		Networks: []*models.PVMInstanceAddNetwork{
+			{
+				NetworkID: &p.serviceConfig.NetworkID,
+			}},
+		Memory:     core.Float64Ptr(p.serviceConfig.Memory),
+		Processors: core.Float64Ptr(p.serviceConfig.Processors),
+		ProcType:   core.StringPtr(p.serviceConfig.ProcessorType),
+		SysType:    p.serviceConfig.SystemType,
+		UserData:   base64.StdEncoding.EncodeToString([]byte(userData)),
+	}
+
+	pvslogger.Printf("CreateInstance: name: %q", instanceName)
+
+	pvsInstances, err := p.powervsService.instanceClient.Create(body)
+	if err != nil {
+		pvslogger.Printf("failed to create an instance : %v", err)
+		return nil, err
+	}
+
+	if len(*pvsInstances) <= 0 {
+		return nil, fmt.Errorf("there are no instances created")
+	}
+
+	ins := (*pvsInstances)[0]
+	instanceID := *ins.PvmInstanceID
+
+	num := 1
+	for {
+		in, err := p.powervsService.instanceClient.Get(*ins.PvmInstanceID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get the instance: %v", err)
+		}
+
+		if *in.Status == "ERROR" {
+			return nil, fmt.Errorf("instance is in error state")
+		}
+
+		if *in.Status == "ACTIVE" {
+			pvslogger.Printf("instance is in desired state: %s", *in.Status)
+			break
+		}
```

Comment:
In the case of IBM Cloud VPC, `CreateInstance` does not wait until the newly created VM insstance becomes ACTIVE state. Instead, `CreateInstance` returns when the API call for `CreateInstance` returns and IP address is assigned.

After `CreateInstance` is called, `cloud-api-adaptor` waits until it can establish a network connection with `agent-protocol-forwarder` running in a new peer pod VM instance.

In the case of PowerVS, are IP addresses assigned to a VM instance before it becomes active? If so, you don't need to wait here.

https://github.com/confidential-containers/cloud-api-adaptor/blob/0935bab6f61988ede05655b9638c3e4371fef847/pkg/adaptor/cloud/ibmcloud/provider.go#L145-L197

https://github.com/confidential-containers/cloud-api-adaptor/blob/0935bab6f61988ede05655b9638c3e4371fef847/pkg/adaptor/cloud/cloud.go#L312-L350

https://github.com/confidential-containers/cloud-api-adaptor/blob/0935bab6f61988ede05655b9638c3e4371fef847/pkg/adaptor/proxy/proxy.go#L214-L216

Reply:
> In the case of PowerVS, are IP addresses assigned to a VM instance before it becomes active? If so, you don't need to wait here.

@yoheiueda, in PowerVS, the IP is assigned to the VM once it is in `ACTIVE` state after the `BUILD` stage(during which it is `nil`) unlike VPC where `0.0.0.0` is assigned at first. Hence the check is needed for fetching the IP and then proceeding with establishing a connection with `agent-protocol-forwarder`.

Reply:
> I think we should have chosen ibmcloudvpc or something for the provider of IBM Cloud VPC. The cloud provider name and build tag name ibmcloud is already referenced in various locations, so changing the name ibmcloud for IBM Cloud VPC for something else from now may causes compatibility issues.

Addressed it, PTAL!

Reply:
> @yoheiueda, in PowerVS, the IP is assigned to the VM once it is in `ACTIVE` state after the `BUILD` stage(during which IP is `nil`) unlike VPC where `0.0.0.0` is assigned at first. Hence the check is needed for fetching the IP and then proceeding with establishing a connection with `agent-protocol-forwarder`.

I see. So, the code looks OK as it is.